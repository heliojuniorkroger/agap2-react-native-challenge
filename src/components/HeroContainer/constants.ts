export const BANNER_HEIGHT = 300;
