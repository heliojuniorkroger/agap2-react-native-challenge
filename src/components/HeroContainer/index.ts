export { HeroContainer } from './HeroContainer';
export type { HeroContainerProps } from './HeroContainer';
