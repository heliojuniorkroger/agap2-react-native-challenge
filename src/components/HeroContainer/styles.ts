import { StyleSheet } from 'react-native';
import { BANNER_HEIGHT } from './constants';

export const styles = StyleSheet.create({
  container: {
    width: '100%',
    height: '100%',
    backgroundColor: '#000814',
  },
  bannerLinearGradient: {
    width: '100%',
    height: BANNER_HEIGHT,
    justifyContent: 'flex-end',
    paddingHorizontal: 20,
    paddingBottom: 15,
  },
  bannerTitle: {
    color: '#fff',
    fontFamily: 'Nunito-Black',
    fontSize: 30,
  },
  bannerDescription: {
    color: '#fff',
    fontFamily: 'Nunito-Regular',
    fontSize: 13,
  },
});
