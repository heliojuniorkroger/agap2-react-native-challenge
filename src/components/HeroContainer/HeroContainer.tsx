import React, { useCallback, useRef } from 'react';
import {
  View,
  Animated,
  ScrollView,
  Text,
  NativeSyntheticEvent,
  NativeScrollEvent,
} from 'react-native';
import LinearGradient from 'react-native-linear-gradient';
import { BANNER_HEIGHT } from './constants';
import { styles } from './styles';

export type HeroContainerProps = {
  title: string;
  description: string;
  imageURI: string;
  children: React.ReactChild;
  scrollViewTestID?: string;
  onScroll?: (event: NativeSyntheticEvent<NativeScrollEvent>) => void;
};

export const HeroContainer = ({
  title,
  description,
  imageURI,
  children,
  scrollViewTestID,
  onScroll: customOnScroll,
}: HeroContainerProps) => {
  const bannerHeightAnim = useRef(new Animated.Value(BANNER_HEIGHT)).current;

  const onScroll = useCallback(
    (event: NativeSyntheticEvent<NativeScrollEvent>) => {
      const { y } = event.nativeEvent.contentOffset;

      if (y < 0) {
        bannerHeightAnim.setValue(BANNER_HEIGHT + Math.abs(y));
      }

      if (typeof customOnScroll === 'function') {
        customOnScroll(event);
      }
    },
    [bannerHeightAnim, customOnScroll],
  );

  return (
    <View style={styles.container}>
      <Animated.Image
        source={{
          uri: imageURI,
        }}
        style={{
          height: bannerHeightAnim,
          width: '100%',
          position: 'absolute',
        }}
      />
      <ScrollView
        testID={scrollViewTestID}
        scrollEventThrottle={1}
        onScroll={onScroll}>
        <LinearGradient
          colors={['transparent', '#000814']}
          style={styles.bannerLinearGradient}>
          <Text style={styles.bannerTitle}>{title}</Text>
          <Text style={styles.bannerDescription}>{description}</Text>
        </LinearGradient>
        {children}
      </ScrollView>
    </View>
  );
};
