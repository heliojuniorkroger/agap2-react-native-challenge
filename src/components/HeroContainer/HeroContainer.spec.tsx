import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import { HeroContainer } from './HeroContainer';

const createDefaultProps = () => ({
  title: 'Hero title',
  description: 'Hero description',
  imageURI: '',
  children: <></>,
  scrollViewTestID: 'scroll-view-test-id',
  onScroll: jest.fn(),
});

describe('testing HeroContainer component', () => {
  it('should match snapshot', () => {
    const props = createDefaultProps();
    const result = render(<HeroContainer {...props} />);
    expect(result).toMatchSnapshot();
  });

  it('should trigger onScroll prop when scrolling the container', async () => {
    const props = createDefaultProps();
    const { getByTestId } = render(<HeroContainer {...props} />);
    const scrollView = getByTestId(props.scrollViewTestID);

    const event = {
      nativeEvent: {
        contentOffset: { y: -15 },
      },
    };

    await fireEvent.scroll(scrollView, event);
    expect(props.onScroll).toBeCalledWith(event);
  });
});
