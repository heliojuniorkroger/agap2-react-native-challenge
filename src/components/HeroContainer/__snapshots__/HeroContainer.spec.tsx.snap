// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing HeroContainer component should match snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#000814",
      "height": "100%",
      "width": "100%",
    }
  }
>
  <Image
    collapsable={false}
    nativeID="animatedComponent"
    source={
      Object {
        "uri": "",
      }
    }
    style={
      Object {
        "height": 300,
        "position": "absolute",
        "width": "100%",
      }
    }
  />
  <RCTScrollView
    onScroll={[Function]}
    scrollEventThrottle={1}
    testID="scroll-view-test-id"
  >
    <View>
      <BVLinearGradient
        colors={
          Array [
            0,
            4278192148,
          ]
        }
        endPoint={
          Object {
            "x": 0.5,
            "y": 1,
          }
        }
        locations={null}
        startPoint={
          Object {
            "x": 0.5,
            "y": 0,
          }
        }
        style={
          Object {
            "height": 300,
            "justifyContent": "flex-end",
            "paddingBottom": 15,
            "paddingHorizontal": 20,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#fff",
              "fontFamily": "Nunito-Black",
              "fontSize": 30,
            }
          }
        >
          Hero title
        </Text>
        <Text
          style={
            Object {
              "color": "#fff",
              "fontFamily": "Nunito-Regular",
              "fontSize": 13,
            }
          }
        >
          Hero description
        </Text>
      </BVLinearGradient>
    </View>
  </RCTScrollView>
</View>
`;
